framework/src/media/media_recorder.c

```diff
@@ -0,0 +1,348 @@
+	if (!g_rc.pcmconf) {
+		RECORD_UNLOCK();
+		return RECORD_ERROR_MALLOC;
+	}
+
+	g_rc.pcmconf->channels = channel;
+	g_rc.pcmconf->rate = sample_rate;
+	g_rc.pcmconf->format = pcm_format;
+	g_rc.format = format;
+
+	RECORD_UNLOCK();
+
+	return RECORD_OK;
+}
+
+record_result_t media_record_prepare(void)
+{
+	RECORD_LOCK();
+
+	if (g_rc.pcmin) {
+		RECORD_UNLOCK();
+		return RECORD_ALREADY_INITIALIZED;
+	}	
+
+	g_rc.pcmin = pcm_open(0, 0, PCM_IN, g_rc.pcmconf);
+	if (!g_rc.pcmin) {
+		RECORD_UNLOCK();
+		return RECORD_ERROR_DRIVER;
+	}
+
+	pcm_prepare(g_rc.pcmin);
+	g_rc.buffer_size = pcm_frames_to_bytes(g_rc.pcmin, pcm_get_buffer_size(g_rc.pcmin));
+
+	RECORD_UNLOCK();
+
+	return RECORD_OK;
+}
+
+bool media_is_recording(void)
+{
+	return (g_rc.state == RECORD_RUNNING);
+}
+
+record_result_t media_stop_record(void)
+{
+	RECORD_LOCK();
+
+	if (g_rc.state == RECORD_PAUSED) {
+		g_rc.state = RECORD_STOPPING;
+		sem_post(&g_rc.sem_pause);
+	} else if (g_rc.state != RECORD_RUNNING) {
+		RECORD_UNLOCK();
+		return RECORD_ERROR_INVALID_STATE;
+	}
+
+	g_rc.state = RECORD_STOPPING;
+	RECORD_UNLOCK();
+
+	return RECORD_OK;
+}
+
+record_result_t media_pause_record(void)
+{
+	RECORD_LOCK();
+
+	if (g_rc.state != RECORD_RUNNING) {
+		RECORD_UNLOCK();		
+		return RECORD_ERROR_INVALID_STATE;
+	}
+
+	g_rc.state = RECORD_PAUSED;
+	RECORD_UNLOCK();
+
+	return RECORD_OK;
+}
+
+record_result_t media_resume_record(void)
+{
+	RECORD_LOCK();
+
+	if (g_rc.state != RECORD_PAUSED) {
+		return RECORD_ERROR_INVALID_STATE;
+	}
+
+	g_rc.state = RECORD_RUNNING;
+	media_record_prepare();	
+	sem_post(&g_rc.sem_pause);
+
+	RECORD_UNLOCK();
+
+	return RECORD_OK;
+}
+
+record_result_t media_record_init(void)
+{
+	if (!b_started) {
+		b_started = true;
+		memset(&g_rc, 0, sizeof(g_rc));
+
+		pthread_mutexattr_init(&g_rc.mutex_attr_record);
+		pthread_mutexattr_settype(&g_rc.mutex_attr_record, PTHREAD_MUTEX_RECURSIVE);
+		pthread_mutex_init(&g_rc.mutex_record, &g_rc.mutex_attr_record);
+
+		g_rc.fd = -1;
+		g_rc.state = RECORD_AVAILABLE;
+		g_rc.pcmin = NULL;
+		g_rc.pcmconf = NULL;
+
+		sem_init(&g_rc.sem_pause, 0, 0);
+	}
+	return RECORD_OK;
+}
+
+static bool write_full(void *ptr, size_t size)
+{
+	int ret;
+	size_t written = 0;
+	while (written < size) {
+		ret = write(g_rc.fd, (char *)ptr + written, size - written);
+		if (ret < 0) {
+			return false;
+		}
+		written += ret;
+	}
+	return true;
+}
+
+int record_and_write(char *buffer)
+{
+	int frames;
+	int size;
+	char inst = ST_PROTO_RECORD_DATA;
+
+	frames = pcm_readi(g_rc.pcmin, buffer, g_rc.buffer_size);
+
+	if (frames > 0) {
+		size = pcm_frames_to_bytes(g_rc.pcmin, frames);
+		if (IS_SOCKET(g_rc.fd))	{
+			write_full(&inst, 1);
+			write_full(&size, sizeof(int));
+		}
+		write_full(buffer, size);
+	}
+	return frames;
+}
+
+int media_recording(void *args)
+{
+	int ret;
+	char *buffer = NULL;
+	char inst = 0;
+	buffer = (char *)malloc(g_rc.buffer_size);
+
+	if (!buffer) {
+		RECORD_LOCK();
+		g_rc.state = RECORD_AVAILABLE;
```

Comment:
RECORD_ERROR????

Reply:
RECORD_AVAILABLE is fine because the recorder is now available to be used by other applications. The state being set here is for future record operations. The result of current operation is returned as malloc error. A mechanism needs to be developed to handle the return value of the pthread.